src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -376,17 +372,69 @@ internal void DoGetExtensionMethods(ArrayBuilder<MethodSymbol> methods, string n
         }
 
 #nullable enable
+        internal static bool IsInvalidExtensionReceiverParameter(ParameterSymbol thisParam)
+        {
+            Debug.Assert(thisParam is not null);
+
+            // For ref and ref-readonly extension members and classic extension methods, receivers need to be of the correct types to be considered in lookup
+            return (thisParam.RefKind == RefKind.Ref && !thisParam.Type.IsValueType) ||
+                (thisParam.RefKind is RefKind.In or RefKind.RefReadOnlyParameter && thisParam.Type.TypeKind != TypeKind.Struct);
+        }
 
-        internal void GetExtensionContainers(ArrayBuilder<NamedTypeSymbol> extensions)
+        internal void GetExtensionMembers(ArrayBuilder<Symbol> members, string? name, string? alternativeName, int arity, LookupOptions options)
```

Comment:
Should we assert that only `LookupOptions.AllMethodsOnArityZero` or `LookupOptions.MustBeInvocableIfMember` are specified?